### DISCUSSION TITLE
Building with 'dotnet publish' in ADO Pipeline

### DISCUSSION THREAD

**Discussion by vikern22** [dotnet/maui#16627] [Q&A] [answered]
### Description

Hi!

Not sure if this is the correct place to post, but;

I'm having some issues building our MAUI-app for iOS in Azure Devops Pipelines. While building for Android seems ok, the iOS build always gets stuck on:

`Optimizing assemblies for size may change the behavior of the app. Be sure to test after publishing. See: https://aka.ms/dotnet-illink
  Optimizing assemblies for size. This process might take a while.`

It builds fine with the same build config in VS for Mac and in terminal on the same Mac - a Mac that seemingly has the same versions as the agent in ADO.

- macOS-13 agent
- Xcode 14.3
- .NET SDK 7.0.306
- MAUI 7.0.92

I've tried using the macOS-12 agent with Xcode 14.2 as well - same result, only with a warning that Xcode 14.3 is recommended.

### Steps to Reproduce

1. Run `dotnet publish -f net7.0-ios -c ReleaseMaui -p:RunTimeIdentifier=ios-arm64 -p:ApplicationVersion=$(VersionOrdinal) -p:ApplicationDisplayVersion=$(VersionString) -p:CodesignKey=$(CodeSignKey) -p:CodesignProvision=${{ parameters.provisioningId }}` in ADO-pipeline
2. Get stuck at optimizing step
3. ADO pipeline gives up and stops the job after 6+ hours

### Version with bug

7.0.92

### Last version that worked well

Unknown/Other

### Affected platforms

iOS

**Comment by vikern22** [comment]
Still struggling with this issue - but don't really know how to approach it. Does anyone have success with building iOS in ADO-pipelines? Is it possible to get more build output, to see what task it actually gets stuck on? Is there any more logical place to post about this?

**Reply by JohnHDev** [reply]
We have .NET MAUI iOS and Android release builds successfully running on ADO. Could you try changing the iOS release config to LLVM off? That seems to ring a bell with me...

**Reply by vikern22** [reply]
Thanks - testing now!

**Reply by vikern22** [reply]
Still stuck with `<EnableLLVM>false</EnableLLVM>` :(

**Comment by vikern22** [comment]
Just tested with a completely fresh "File -> new"-project, same issue occurs in ADO pipelines

**Reply by JohnHDev** [reply]
can you share your pipeline yaml (excluding confidential project/solution names etc), I can compare to ours.

**Reply by vikern22** [reply]
Thanks for your time!

jobs:
  - job: BuildJob
    displayName: 'Build MAUI(Test)'
    steps:
      - task: Bash@3
        displayName: Install MAUI Workload
        inputs:
          targetType: 'inline'
          script: |
            dotnet nuget locals all --clear
            dotnet workload install maui --source https://pkgs.dev.azure.com/dnceng/public/_packaging/dotnet7/nuget/v3/index.json --source https://api.nuget.org/v3/index.json

      - task: DotNetCoreCLI@2
        displayName: 'Restore packages'
        inputs:
          command: 'restore'
          feedsToUse: 'select'
          vstsFeed: ${{ parameters.feed }}
          projects: 'Source/Mobil/TestMauiBuild/TestMauiBuild.sln'

      - task: InstallAppleCertificate@2
        displayName: 'Install an Apple certificate'
        inputs:
          certSecureFile: ${{ parameters.certificateFile }}
          certPwd: ${{ parameters.certificatePassword}}
          setUpPartitionIdACLForPrivateKey: false
          keychain: 'iPhone Distribution'

      - task: InstallAppleProvisioningProfile@1
        displayName: 'Install an Apple provisioning profile'
        inputs:
          provProfileSecureFile: ${{ parameters.provisioningFile }}

      - bash: |
          echo '##vso[task.setvariable variable=MD_APPLE_SDK_ROOT;]'/Applications/Xcode_14.3.app;sudo xcode-select --switch /Applications/Xcode_14.3.app/Contents/Developer   
        displayName: 'Use specific Xcode version'

      - task: Bash@3
        displayName: Build iOS App
        inputs:
          targetType: 'inline'
          script: |
            cd Source/Mobil/TestMauiBuild
            dotnet publish -f net7.0-ios -c Release -p:RunTimeIdentifier=ios-arm64 -p:CodesignKey=${{ parameters.codeSignKey }} -p:CodesignProvision=${{ parameters.provisioningId }} -p:ArchiveOnBuild=true

**Reply by JohnHDev** [reply]
A few differences I spotted:
We install the cert and profile first (shouldn't maker any difference)

We install maui with this on CmdLine@2:
    sudo dotnet workload install maui

We switch to Xcode 14.1, probably due to an issue where Maui didn't yet support Xcode 14.3 which has since been resolved I believe.

Our dotnet publish is different too, mainly the BuildIpa parameter:
            dotnet publish -f net7.0-ios -c RELEASE /p:ApplicationDisplayVersion='$(major).$(minor).$(patch)' /p:ApplicationVersion=${REV} /p:BuildIpa=True /p:RuntimeIdentifier=ios-arm64 "/p:CodesignKey=$(APPLE_CERTIFICATE_SIGNING_IDENTITY)" "/p:CodesignProvision=$(APPLE_PROV_PROFILE_UUID)"

We then have a deploy to TestFlight step, but you aren't getting that far.

A few things to try at least, hope it helps.

**Reply by vikern22** [reply]
I'll give it a shot. Thanks!

**Reply by vikern22** [reply]
No luck :/

**Reply by JohnHDev** [reply]
Recreate provisioning profile and cert? Without logging in and taking a look, I doubt I can be much help now. I have just run a release build on macos13 and it is working for us.

**Reply by vikern22** [reply, marked as answer]
Seemed to be an issue with the certificate indeed. We tested building on a Mac that just had the certificate installed - and on the optimizing step a password prompt appeared!

We then discovered that we were trying to install the certificate to a specific keychain on the agent, and changed it to use the default, and now it runs to completion.

**Reply by alexcoupalNOV** [reply]
Hi @vikern22 , your finding saved me. I would have never find that one on my own. 
Could you please tell me how you changed it to use the default keychain ?
Thanks
